### PR TITLE
Major refactoring

### PR DESCRIPTION
**Pull request by withrocks**
- Cleanup and refactoring (tech debt form the prototype stage.)
- Remove logic from the ExtensionContext into different services. This
  allows us to only test the services.
- Move several files around, e.g. ResultFile to domain/result_file.py
- Factory method for the ExtensionContext.
- Create a repository for local and remote files, which can be
  injected into services
- Major changes to the StepRepository. It now returns artifacts in pairs
  (input, output) rather than one list of inputs and another of outputs.
- Add the ArtifactService, which filters data from the step repository.
  Services are designed to be tested while only their underlying repos
  should be mocked. This way mocking is limited data only and unit tests
  will cover most of the logic.
- Move code to the new FileService. This handles downloading files and
  parsing them to XML. Some of this code was in the ExtensionContext
  before.
- Add tests, coverage now at approx 92% when integration tests are also run
- Add a Jupyter notebook for a quick overview of how things work together.

### CODE REVIEW

File: clarity_ext/context.py [hunk 1]
```diff
@@ -1,180 +1,127 @@
...
     """
-    def __init__(self, session, step_repository, cache=False, logger=None):
+    def __init__(self, session, artifact_service, file_service, cache=False, logger=None):
         """
         Initializes the context.
 
         :param session: An object encapsulating the connection to Clarity
-        :param step_repository: A repository for accessing data from the current step
+        :param artifact_service: Provides access to artifacts in the current step
+        :param file_service: Provides access to result files locally on the machine.
```
Comment (withrocks) [inline clarity_ext/context.py]: Fix help string: File service has to do with both local and remote files.

File: clarity_ext/context.py [hunk 2]
```diff
@@ -1,180 +1,127 @@
...
-        resource = [f for f in self.output_result_files if f.id == id][0]
-        return ResultFile(resource, self.units)
+        return self.artifact_service.output_file_by_id(file_id)
 
     @property
     def output_result_files(self):
-        for _, output in self.current_step.input_output_maps:
-            if output["output-type"] == "ResultFile":
-                yield output["uri"]
+        # TODO: This does not belong here anymore, move to a lower level
```
Comment (withrocks) [inline clarity_ext/context.py]: Remove TODO

File: clarity_ext/domain/container.py [hunk 3]
```diff
@@ -10,26 +11,26 @@ class Well(DomainObjectMixin):
...
+        return "{}:{}".format(self.position.row, self.position.col)
 
     @property
     def index_down_first(self):
         # The position is 1-indexed
-        return (self.position.col - 1) * self.plate.size.height + self.position.row
+        return (self.position.col - 1) * self.container.size.height + self.position.row
 
 
 class PlatePosition(namedtuple("PlatePosition", ["row", "col"])):
```
Comment (withrocks) [inline clarity_ext/domain/container.py]: Edvard: Change name to ContainerPosition

File: clarity_ext/repository/container_repository.py [hunk 4]
```diff
@@ -0,0 +1,20 @@
+from clarity_ext.domain import Container
+
+
+class ContainerRepository:
```
Comment (withrocks) [inline clarity_ext/repository/container_repository.py]: Rename to service?

File: clarity_ext/repository/step_repository.py [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+            artifact_keys.add(input["uri"])
+            artifact_keys.add(output["uri"])
+        artifacts = self.session.api.get_batch(artifact_keys)
+        artifacts_by_uri = {artifact.uri: artifact for artifact in artifacts}
+        for input, output in input_output_maps:
+            input['uri'] = artifacts_by_uri[input['uri'].uri]
+            output['uri'] = artifacts_by_uri[output['uri'].uri]
+
+        ret = []
+        container_repo = ContainerRepository()
```
Comment (withrocks) [inline clarity_ext/repository/step_repository.py]: Look into if this uses get_batch too (I don't think it does).

File: clarity_ext/service/artifact_service.py [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+
+    def all_analyte_pairs(self):
+        """
+        Returns all analytes in a step as an artifact pair (input/output)
+        """
+        # TODO: Ensure that the repo returns artifacts sorted, i.e. that the output really matches
+        # the input
+        pairs = self.step_repository.all_artifacts()
+
+        # TODO: Unique needed. check all usages BEFORE CHECKIN?
```
Comment (withrocks) [inline clarity_ext/service/artifact_service.py]: Whitespace

Comment (withrocks) [inline clarity_ext/service/artifact_service.py]: Remove TODO

File: test/integration/domain/test_result_file.py [hunk 7]
```diff
@@ -0,0 +1,67 @@
+import requests_cache
+import unittest
+from clarity_ext.domain import Container, ResultFile
+from clarity_ext.context import ExtensionContext
+import os
+
+requests_cache.install_cache(
```
Comment (withrocks) [inline test/integration/domain/test_result_file.py]: Remove this

**Comment by GitEdvard** [comment]
Ok

**Pull request merged**